### ISSUE TITLE
Implementation of Group equivariant convolutions

### ISSUE THREAD

**Issue by sairaamVenkatraman** [pytorch/pytorch#21411]
## 🚀 Feature
An implementation of group-equivariant convolutions defined by Cohen et. al in the paper 'Group equivariant convolutions' - in ICML 2016.
Specifically, an implementation of the group-convolution over the group of translations and rotations of multiples of 90 degrees at any centre, and the same group extended by reflections.  

## Motivation
Group-equivariant convolutions (g-cnns) extend the equivariance of convolutions to beyond translations. Cohen et. al argue in their paper that the parameter-sharing in g-cnns allow for more efficient use of parameters.
Further, limited experiments conducted by Adam Bielski in his github repository https://github.com/adambielski/pytorch-gconv-experiments, suggest that g-cnns improve upon the performance of regular cnn networks in at least some cases.
I'm not sure if the existing pytorch implementations are efficient as they take considerably more time to run than nn.Conv2d. 

To sum up, it would be awesome if you guys could create nn.GConv2d() for group-equivariant convolutions on at least the two groups mentioned before.

## Additional context
TS cohen's original implementation of group-equivariant convolutional networks: https://github.com/tscohen/GrouPy

**Comment by cpuhrsch** [comment]
@sairaamVenkatraman could you expand a bit more on the downsides of the current implementation mentioned in the external repositories? Is this mostly about speed?

**Comment by sairaamVenkatraman** [comment]
@cpuhrsch My isssues with the current implementation are mainly about speed. I'm not sure if that is intrinsic, or if its due some inefficiency in their implementation. I'll try and post some timing comparisons, maybe, when I can.

**Comment by TDHTTTT** [comment]
Is this still an interesting issue for pytorch? If so, I am willing to contribute.
